pandas/core/window/numba_.py

```diff
@@ -0,0 +1,110 @@
+import types
+from typing import Callable, Dict, Optional, Tuple
+
+import numpy as np
+
+from pandas.compat._optional import import_optional_dependency
+
+
+def make_rolling_apply(func, args, nogil, parallel, nopython):
+    numba = import_optional_dependency("numba")
+
+    if parallel:
+        loop_range = numba.prange
+    else:
+        loop_range = range
+
+    if isinstance(func, numba.targets.registry.CPUDispatcher):
+        # Don't jit a user passed jitted function
+        numba_func = func
+    else:
+
+        @numba.generated_jit(nopython=nopython, nogil=nogil, parallel=parallel)
+        def numba_func(window, *_args):
+            if getattr(np, func.__name__, False) is func or isinstance(
+                func, types.BuiltinFunctionType
+            ):
+
+                def impl(window, *_args):
+                    return func(window, *_args)
+
+                return impl
+            else:
+                jf = numba.jit(func, nopython=nopython)
+
+                def impl(window, *_args):
+                    return jf(window, *_args)
+
+                return impl
+
+    @numba.jit(nopython=nopython, nogil=nogil, parallel=parallel)
+    def roll_apply(
+        values: np.ndarray, begin: np.ndarray, end: np.ndarray, minimum_periods: int,
+    ):
+        result = np.empty(len(begin))
+        for i in loop_range(len(result)):
+            start = begin[i]
+            stop = end[i]
+            window = values[start:stop]
+            count_nan = np.sum(np.isnan(window))
+            if len(window) - count_nan >= minimum_periods:
+                result[i] = numba_func(window, *args)
+            else:
+                result[i] = np.nan
+        return result
+
+    return roll_apply
+
+
+def generate_numba_apply_func(
+    args: Tuple,
+    kwargs: Dict,
+    func: Callable,
```

Comment:
Does `Callable` need to return an `ndarray`? Adding sub-types to `Callable` helps immensely so would be preferable to add if so

Reply:
`Callable` here should take in an ndarray and potentially `*args` and return a scalar. What would be the best way to type that?

Reply:
Hmm not sure about the *args piece but here are the docs for it:

https://docs.python.org/3/library/typing.html#typing.Callable

So I would think `Callable[[np.ndarray, ...], Scalar]` if it works otherwise `Callable[..., Scalar]` is the best we can do.

Scalar can be imported from `pandas._typing`

Reply:
Thanks. I'll give that a shot.

Reply:
Had to go with `Callable[..., Scalar]` as also specifying `np.ndarray` was raising errors.